libexec/python/tests/test_core.py

```diff
@@ -149,18 +149,18 @@ def test_parse_image_uri(self):
         self.assertTrue(digest['namespace'] == 'mix')
 
         print("Case 8: Custom uri should use it.")
-        image_name = "catdog://meow/mix/%s:%s" % (self.repo_name,
-                                                  self.tag)
+        image_name = "catdog://meow/mix/tenders/%s:%s" % (self.repo_name,
+                                                          self.tag)
         digest = parse_image_uri(image_name, uri="catdog://")
         self.assertTrue(digest['registry'] == 'meow')
-        self.assertTrue(digest['namespace'] == 'mix')
+        self.assertTrue(digest['namespace'] == 'mix/tenders')
 
         print("Case 9: Digest version should be parsed")
-        image_name = ("catdog://meow/mix/%s:%s@sha:256xxxxxxxxxxxxxxx"
+        image_name = ("catdog://meow/mix/original/choice/%s:%s@sha:256xxxxxxxxxxxxxxx"  # noqa
                       % (self.repo_name, self.tag))
         digest = parse_image_uri(image_name, uri="catdog://")
         self.assertTrue(digest['registry'] == 'meow')
-        self.assertTrue(digest['namespace'] == 'mix')
+        self.assertTrue(digest['namespace'] == 'mix/original/choice')
```

Comment:
I've inherited this, not created it :)

My basic problem is that the existing code doesn't deal with multiple levels in the path.  I'm not particularly well equipped to argue about whether the extra bits are registry or namespace....

We have things like this:

```
docker pull https://foo.blah.com:5001/myorg/aproject/repo_name@sha256:530b5227f92f870aff05c040c5b1f27c15b5978dbf7704e877b71691ed58ede2
```

In this case I *think* that `foo.blah.com:5001` is the registry.

In your example above, I don't think that `catdog` is a uri, I think that the entire thing is a URI and that `catdog` is the scheme (based on e.g. [Wikipedia's URI syntax example](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax)).

Stealing your neat labeling trick, I'd end up with:

```
https://foo.blah.com:5001/myorg/aproject/repo_name@sha256:530b5227f92f870aff05c040c5b1f27c15b5978dbf7704e877b71691ed58ede2
[scm]://[    registry   ][  namespace  ][  repo  ]@[digest]
```

Part of my motivation for labelling it this way is we have another "registry" at `https://foo.blah.com:5002/`.

On [the gripping hand](https://en.wikipedia.org/wiki/The_Gripping_Hand), perhaps it's all registries and repositories and "namespaces" are just FigNewtons of our imagination?

```
[vagrant@localhost singularity]$ http http://blah.moo.com:5000/v2/_catalog
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Date: Sat, 12 Aug 2017 23:46:31 GMT
Docker-Distribution-Api-Version: registry/2.0
Transfer-Encoding: chunked
X-Content-Type-Options: nosniff

{
    "repositories": [
        "demo/blah-map",
        "moose/foo",
        "moose/poodle",
        "hartzelg/doh",
        "hartzelg/hello-world",
        "hartzelg/scooby-doo",
        "hartzelg-this-one-too",
        "hartzelg-hello-world",
        "userx/tuna",
        "research/sloopy/doh-bar",
        "research/sloopy/pete-moose",
[...]
```